### PR TITLE
fashion_mnist colab examples

### PR DESCRIPTION
**Pull request by sgwhat**
1. create pytorch_distributed version
2. create bigdl version
3. add comments
4. support tensorbord for loss descent visualization

### CODE REVIEW

**Comment by sgwhat** [comment]
Still need to add test script and remove original python file

**Comment by jason-dai** [comment]
Why do we have two notebooks? Just need one for colab.

**Comment by hkvision** [comment]
> Why do we have two notebooks? Just need one for colab.

It may be not clear to put two backends in one notebook? You prefer to keep only one backend for demo or put two together?

**Comment by jason-dai** [comment]
> > Why do we have two notebooks? Just need one for colab.
> 
> It may be not clear to put two backends in one notebook? You prefer to keep only one backend for demo or put two together?

Just pick one backend.

**Comment by hkvision** [comment]
> > > Why do we have two notebooks? Just need one for colab.
> > 
> > 
> > It may be not clear to put two backends in one notebook? You prefer to keep only one backend for demo or put two together?
> 
> Just pick one backend.

Okay. But since we have already done both, can we just keep them in our repo and only expose one on our documentation website?

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb [hunk 1]
```diff
@@ -0,0 +1,702 @@
...
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "e_jJ1tnZuMJZ"
+      },
+      "source": [
+        "import argparse\n",
+        "\n",
+        "parser = argparse.ArgumentParser(description='PyTorch Tensorboard Example')\n",
+        "parser.add_argument('--cluster_mode', type=str, default=\"local\",\n",
```
Comment (Le-Zheng) [inline docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:221]: argparse may not work in ipynb. Just remove this cell.

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "fhQsdAhLEQxa"
+      },
+      "source": [
+        "# move into logdata file for tensorboard using\n",
+        "%tensorboard --logdir '/content/runs/fashion_mnist_experiment_1'"
```
Comment (Le-Zheng) [inline docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:674]: why not display the result?

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb (hunk 1 above)
Comment (sgwhat) [reply docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:221]: it could work well actually, I modified some code to run argparse cell successfully in colab.

Comment (sgwhat) [reply docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:221]: I put this cell when two backends were included in one notebook, seems it's useless now, I'll delete this cell.

**Comment by jason-dai** [comment]
> > > > Why do we have two notebooks? Just need one for colab.
> > > 
> > > 
> > > It may be not clear to put two backends in one notebook? You prefer to keep only one backend for demo or put two together?
> > 
> > 
> > Just pick one backend.
> 
> Okay. But since we have already done both, can we just keep them in our repo and only expose one on our documentation website?

It's a bad idea to have redundant codes; just add one for colab.

**Comment by sgwhat** [comment]
> > > > > Why do we have two notebooks? Just need one for colab.
> > > > 
> > > > 
> > > > It may be not clear to put two backends in one notebook? You prefer to keep only one backend for demo or put two together?
> > > 
> > > 
> > > Just pick one backend.
> > 
> > 
> > Okay. But since we have already done both, can we just keep them in our repo and only expose one on our documentation website?
> 
> It's a bad idea to have redundant codes; just add one for colab.

Sure, I will keep only one notebook.

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "tUGTC8lT07ZE"
+      },
+      "source": [
+        "# Install latest pre-release version of Analytics Zoo \n",
+        "# Installing Analytics Zoo from pip will automatically install pyspark, bigdl, and their dependencies.\n",
+        "!pip install --pre --upgrade analytics-zoo[ray]"
```
Comment (hkvision) [inline docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:154]: for bigdl backend, no need [ray]

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "2EW8lPzz0_i0"
+      },
+      "source": [
+        "# Install python dependencies\n",
+        "!pip install torch==1.7.1 torchvision==0.8.2\n",
+        "!pip install -U ray\n",
```
Comment (hkvision) [inline docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:167]: remove this line

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+        "!pip install jep==3.9.0\n",
+        "!pip install six cloudpickle\n",
```
Comment (hkvision) [inline docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:169]: https://github.com/intel-analytics/BigDL/blob/master/pyspark/setup.py#L92
BigDL will already install six as a dependency, so we can remove six here and also in the README of the original example?

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb [hunk 6]
```diff
@@ -0,0 +1,647 @@
...
+      "metadata": {
+        "id": "D-mCzxMaN-Fc"
+      },
+      "source": [
+        "#import necessary libraries and modules\n",
+        "from __future__ import print_function\n",
+        "import numpy as np\n",
+        "\n",
+        "from zoo.orca import init_orca_context, stop_orca_context\n",
+        "from zoo.orca import OrcaContext"
```
Comment (hkvision) [inline docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:198]: merge these two imports into one?

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb [hunk 7]
```diff
@@ -0,0 +1,647 @@
...
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "jipxQZNW_0c6"
+      },
+      "source": [
+        "After defining your model, you need to define a *Model Creator Function* that returns an instance of your model, and a *Optimizer Creator Function* that returns a pytorch optimizer."
```
Comment (hkvision) [inline docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:297]: no need creator for bigdl backend

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb [hunk 8]
```diff
@@ -0,0 +1,647 @@
...
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "N0ZmT18t5C_6"
+      },
+      "source": [
+        "A brief overview of the dashboards shown (tabs in top navigation bar):\n",
```
Comment (hkvision) [inline docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:429]: refactor step 4 to make it clearer.

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb [hunk 9]
```diff
@@ -0,0 +1,647 @@
...
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "Xp8hzltsZCQP"
+      },
+      "source": [
+        "orca_estimator.set_tensorboard(tensorboard_dir, \"bigdl\")"
```
Comment (hkvision) [inline docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:541]: add a comment what "bigdl" argument means? 

tensorboard_dir = "runs"
writer = SummaryWriter(tensorboard_dir + '/bigdl')

Why in writer we have bigdl but here we still need bigdl?

File: docs/docs/colab-notebook/run-notebook-pytorch.sh [hunk 10]
```diff
@@ -79,3 +79,40 @@ do
 	echo "Complete #${index} with time ${time} seconds"	
 	index=$((index+1))
 done
+
+# orca example test
+echo "orca examples test start"
+
+dir=${ANALYTICS_ZOO_HOME}/docs/docs/colab-notebook/orca/examples
+pytorchFiles=("fashion_mnist_pytorch_distributed" "fashion_mnist_bigdl")
```
Comment (hkvision) [inline docs/docs/colab-notebook/run-notebook-pytorch.sh:87]: only keep and test bigdl

File: docs/docs/colab-notebook/run-notebook-pytorch.sh [hunk 11]
```diff
@@ -79,3 +79,40 @@ do
...
+	sed -i "s/import os/#import os/g" ${filename}.py
+	sed -i "s/import sys/#import sys/g" ${filename}.py
+	sed -i 's/^[^#].*environ*/#&/g' ${filename}.py
+	sed -i 's/^[^#].*__future__ */#&/g' ${filename}.py
+	sed -i "s/_ = (sys.path/#_ = (sys.path/g" ${filename}.py
+	sed -i "s/.append/#.append/g" ${filename}.py
+	sed -i 's/^[^#].*site-packages*/#&/g' ${filename}.py
+	sed -i 's/version_info/#version_info/g' ${filename}.py
+	sed -i 's/python_version/#python_version/g' ${filename}.py
+
```
Comment (hkvision) [inline docs/docs/colab-notebook/run-notebook-pytorch.sh:111]: can we set epochs =1 during the test?

File: docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb (hunk 9 above)
Comment (sgwhat) [reply docs/docs/colab-notebook/orca/examples/fashion_mnist_bigdl.ipynb:541]: bigdl is the path to store image and graph files in the writer, it's also the app_name to save the training log file.

The same name could display scalars, images and graphs in the same tensorboard.

**Review by hkvision — APPROVED** [review]

**Comment by hkvision** [comment]
@jason-dai We will merge this first. We would update readthedocs after several notebooks are finished.

**Comment by sgwhat** [comment]
jenkins passed http://10.239.176.111:18888/job/ZOO-PR-Quickstart-notebook/
